Add alerts infra and basic alert CnaoDown

RamLavi opened the pull request:
**What this PR does / why we need it**:
This PR add infra support and basic alert `CnaoDown` to CNAO 

**Special notes for your reviewer**:
manually tested with CRC:
metric:
![monitoring-graph](https://user-images.githubusercontent.com/48101085/126952703-353344e5-f19f-4b74-82e9-16ba3511e5dd.jpeg)
you can see that CNAO being down for >5 minutes is casuing the alert below
alert:
![alarm fired](https://user-images.githubusercontent.com/48101085/126952736-d0799238-699d-4971-a0bc-3b884ae8e551.jpeg)

**Release note**:

```release-note
NONE
```

RamLavi commented:
/hold
this PR depends on PR #923 to be merged.

RamLavi commented on cmd/manager/main.go:
```diff
@@ -72,7 +72,7 @@ func main() {
 	mgr, err := manager.New(cfg, manager.Options{
 		Namespace:          namespace,
-		MetricsBindAddress: fmt.Sprintf("%s:%d", metricsHost, metricsPort),
+		MetricsBindAddress: fmt.Sprintf("%s:%d", operatorSdkMetricsHost, operatorSdkHttpMetricsPort),
```
I'm not certain as to if I can/should use the same port as operatorSDK (8383), 
when I tried it failed on "address already set".
if so then I'm not sure if the which port value does `MetricsBindAddress` should contain

RamLavi replied:
I ended up removing the operator-sdk metric altogether..  the way I see it, if we don't use them,  we shouldn't keep them

RamLavi replied:
update - brought it back :)

RamLavi commented:
/hold cancel 
#923 is merged, current PR rebased.

will still keep the PR under WIP until I check this under crc with prometheus deployed

RamLavi commented:
/hold
relying on #942 bug fix.

RamLavi commented on pkg/monitoring/metrics.go:
```diff
@@ -0,0 +1,30 @@
+			Handler: mux,
+		}
+		log.Printf("Starting Prometheus metrics endpoint server")
+		err := server.ListenAndServe()
```
@phoracek  This is still an open issue whether we must use TLS..
pending getting a response from OCP monitoring slack,

RamLavi replied:
According to OCP monitoring slack, we're good.

RamLavi commented:
/retest
ci hiccup

RamLavi commented:
/retest

RamLavi commented:
/hold cancel

RamLavi commented:
/hold
we still need to get an answer egarding the TLS communication

qinqon commented:
/hold 

Since we are using controller-runtime here and it has already a metrics capacity I think it would be better to ingrate it 

https://book.kubebuilder.io/reference/metrics.html

RamLavi commented:
[here](https://github.com/kubevirt/cluster-network-addons-operator/compare/14effb9256144e426ae25b3e4bd7b7d8cbbb9390..a62d445754c22b0b21c749827e0b9d1350ac2e4b) I changed the condition to deploy monitoring manifests to only checking relevant CRDs, and also added another env-var for the serviceAccount name

RamLavi commented:
> /hold
> 
> Since we are using controller-runtime here and it has already a metrics capacity I think it would be better to ingrate it
> 
> https://book.kubebuilder.io/reference/metrics.html

fixed [here](https://github.com/kubevirt/cluster-network-addons-operator/compare/a62d445754c22b0b21c749827e0b9d1350ac2e4b..6bf3b770b245c610dd81b4da47003d185c7c7e50)

oshoval commented on cluster/up.sh:
```diff
@@ -18,6 +18,13 @@ set -ex
+
+export KUBEVIRT_DEPLOY_PROMETHEUS=true
+if [[ ($KUBEVIRT_PROVIDER =~ k8s-1\.1.*) || ($KUBEVIRT_PROVIDER =~ k8s-1.20) ]]; then
+	  echo "Prometheus is only supported for providers >= k8s-1.21\n Disabling Prometheus"
```
nit: ident

RamLavi replied:
thanks! [Done](https://github.com/kubevirt/cluster-network-addons-operator/compare/7a61ef72f51fcc207386bac49d2e59627e0349b2..cacead5ccc2b09349d5ce311de4f84314d3e6339)

qinqon commented on cluster/up.sh:
```diff
@@ -18,6 +18,13 @@ set -ex
 SCRIPTS_PATH="$(dirname "$(realpath "$0")")"
 source ${SCRIPTS_PATH}/cluster.sh
+
+export KUBEVIRT_DEPLOY_PROMETHEUS=true
+if [[ ($KUBEVIRT_PROVIDER =~ k8s-1\.1.*) || ($KUBEVIRT_PROVIDER =~ k8s-1.20) ]]; then
+    echo "Prometheus is only supported for providers >= k8s-1.21\n Disabling Prometheus"
+    export KUBEVIRT_DEPLOY_PROMETHEUS=false
+fi
+
```
We should do this at kubevirtci right ?

RamLavi replied:
removed this condition. 
To be clear - this means the from this commit onward `make cluster-up` with provider<1.21 will fail on kubevirtci (which is ok, the error message there is quite clear)

qinqon commented on cmd/manager/main.go:
```diff
@@ -72,7 +62,7 @@ func main() {
 	mgr, err := manager.New(cfg, manager.Options{
 		Namespace:          namespace,
-		MetricsBindAddress: fmt.Sprintf("%s:%d", metricsHost, metricsPort),
+		MetricsBindAddress: monitoring.GetMetricsAddress(),
```
We are not adding an argument to change metrics bind address from the container and default values https://github.com/kubernetes-sigs/controller-runtime/blob/b704f447ea7c8f7059c6665143a4aa1f6da28328/pkg/metrics/listener.go#L30 are good enough, let's just not set this so we use the default metrics bind address.

RamLavi replied:
not sure I got you. why do we not want to use a specific port, that we can also reference to on components.go?

RamLavi replied:
ok I change `monitoring.GetMetricsAddress()` to use `metrics.DefaultBindAddress` , is that alright?

qinqon replied:
We don't need to pass anything and we don't need "GetMetricsAddress" since we are using controller-runtime's defaults.

RamLavi replied:
I can  remove it, but then how do you feel about the fact that we're already parsing the port on `GetMetricsPort` using  `DefaultBindAddress` ([here](https://github.com/kubevirt/cluster-network-addons-operator/pull/928/files#diff-abbf397a1a9176a1fd8d85bc268d3acf4d55fb789feabd1c2ae818395e26083eR69))?

My thinking was that if we're already using DefaultBindAddress exclicitly there then I should be consistent.. waiting to see what you say

qinqon commented on data/monitoring/service.yaml:
```diff
@@ -0,0 +1,17 @@
+spec:
+  ports:
+    - name: metrics
+      port: 443
```
If we are using the default 8080 would be good also since we are not using TLS 443 does not make sense here.

qinqon commented on pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go:
```diff
@@ -598,6 +605,25 @@ func isSCCAvailable(c kubernetes.Interface) (bool, error) {
 	return isResourceAvailable(c, "securitycontextconstraints", "security.openshift.io", "v1")
 }
 
+func isMonitoringAvailable(c kubernetes.Interface) (bool, error) {
+	prometheusRuleResourceAvailable, err := isResourceAvailable(c, "customresourcedefinitions/prometheusrules.monitoring.coreos.com", "apiextensions.k8s.io", "v1")
+	if err != nil {
+		return false, errors.Wrap(err, "failed to check if prometheusRule resource is available")
+	}
+
+	serviceMonitorResourceAvailable, err := isResourceAvailable(c, "customresourcedefinitions/servicemonitors.monitoring.coreos.com", "apiextensions.k8s.io", "v1")
+	if err != nil {
+		return false, errors.Wrap(err, "failed to check if serviceMonitor resource is available")
+	}
+
+	if prometheusRuleResourceAvailable && serviceMonitorResourceAvailable {
+		return true, nil
+	}
+
+	log.Printf("will not deploy monitoring manifests: not all monitoring resources are available: %s: %v, %s, %v", monitoringv1.PrometheusRuleKind, prometheusRuleResourceAvailable, monitoringv1.ServiceMonitorsKind, serviceMonitorResourceAvailable)
+	return false, nil
+}
+
```
Return false if there is an error but print the error we don't want CNAO to fail only if monitoring cannot be activated.

RamLavi replied:
I will change in the function calling this on the controller so that it will not fail and just log it.

qinqon commented on pkg/monitoring/metrics.go:
```diff
@@ -18,3 +21,21 @@ func init() {
 	return fmt.Sprintf("%s:%s", monitoringAddress, monitoringPort)
 }
+
+func GetMonitoringNamespace() string {
+	monitoringNamespaceFromEnv := os.Getenv("MONITORING_NAMESPACE")
+
+	if monitoringNamespaceFromEnv != "" {
+		return monitoringNamespaceFromEnv
+	}
+	return defaultMonitoringNamespace
+}
+
+func GetMonitoringServiceAccount() string {
+	monitoringServiceAccountFromEnv := os.Getenv("MONITORING_SERVICE_ACCOUNT")
+
+	if monitoringServiceAccountFromEnv != "" {
+		return monitoringServiceAccountFromEnv
+	}
+	return defaultServiceAccountName
+}
```
We are already at "monitoring" package we can do `monitoring.GetNamespace` `monitorig.GetServiceAccount`

RamLavi replied:
Done

qinqon commented on pkg/network/monitoring.go:
```diff
@@ -0,0 +1,31 @@
+	"github.com/kubevirt/cluster-network-addons-operator/pkg/render"
+)
+
+func renderMonitoring(manifestDir string, clusterInfo *ClusterInfo) ([]*unstructured.Unstructured, error) {
+	if !clusterInfo.MonitoringAvailable {
+		return nil, nil
+	}
+
+	// render the manifests on disk
+	data := render.MakeRenderData()
+	data.Data["Namespace"] = os.Getenv("OPERAND_NAMESPACE")
+	data.Data["MonitoringNamespace"] = monitoring.GetMonitoringNamespace()
+	data.Data["MonitoringServiceAccount"] = monitoring.GetMonitoringServiceAccount()
+
+	objs, err := render.RenderDir(filepath.Join(manifestDir, "monitoring"), &data)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to render monitoring manifests")
+	}
+
+	return objs, nil
+}
```
This has nothing to do with "network" package, it belongs to "monitoring" package so we do `monitoring.Render`

RamLavi replied:
well the entire Render function is in `network.go`
I moved it to `pkg/monitoring/metrics.go`, but not sure it is comfy here.. 
what do you say?

qinqon commented on test/check/check.go:
```diff
@@ -120,6 +122,36 @@ func PlacementListFromComponentDaemonSets(component Component) ([]cnao.Placement
 	return placementList, nil
 }
 
+func getMonitoringNamespace() (string, error) {
```
Why don't we use the production fruncion ?

RamLavi replied:
I see no reason why not :)

qinqon commented on test/check/check.go:
```diff
@@ -120,6 +122,36 @@ func PlacementListFromComponentDaemonSets(component Component) ([]cnao.Placement
+	}
+
+	ns := corev1.Namespace{}
+	err = framework.Global.Client.Get(context.Background(), types.NamespacedName{Name: monitoringNamespace}, &ns)
+	if err == nil {
+		return true, nil
+	} else if apierrors.IsNotFound(err) {
+		return false, nil
+	}
+
+	return false, err
```
We have to make this more idiomatic doing some kind of fail fast.

```golang
if err != nil {
  if apierrors.IsNotFound(err) {
    return false, nil
  }
  return false, err
}
return false, nil
```

RamLavi replied:
removed the function entirely in favor on the production function

qinqon commented on test/check/check.go:
```diff
@@ -245,6 +277,21 @@ func CheckForLeftoverObjects(currentVersion string) {
 	Expect(serviceAccounts.Items).To(BeEmpty(), "Found leftover objects from the previous operator version")
+
+	services := corev1.ServiceList{}
+	err = framework.Global.Client.List(context.Background(), &services, &listOptions)
+	Expect(err).NotTo(HaveOccurred())
```
Expect(framework.Global.Client.List(context.Background(), &services, &listOptions)).To(Succees())

RamLavi replied:
done

qinqon commented on test/check/check.go:
```diff
@@ -245,6 +277,21 @@ func CheckForLeftoverObjects(currentVersion string) {
+	Expect(services.Items).To(BeEmpty(), "Found leftover objects from the previous operator version")
+
+	serviceMonitors := monitoringv1.ServiceMonitorList{}
+	err = framework.Global.Client.List(context.Background(), &serviceMonitors, &listOptions)
```
ditto

RamLavi replied:
done

qinqon commented on test/check/check.go:
```diff
@@ -245,6 +277,21 @@ func CheckForLeftoverObjects(currentVersion string) {
+	Expect(serviceMonitors.Items).To(BeEmpty(), "Found leftover objects from the previous operator version")
+
+	prometheusRules := monitoringv1.PrometheusRuleList{}
+	err = framework.Global.Client.List(context.Background(), &prometheusRules, &listOptions)
```
ditto

RamLavi replied:
done

qinqon commented on test/e2e/lifecycle/main_test.go:
```diff
@@ -39,6 +40,11 @@ var _ = BeforeSuite(func() {
 	err := framework.AddToFrameworkScheme(apis.AddToScheme, &cnaov1.NetworkAddonsConfigList{})
 	Expect(err).ToNot(HaveOccurred())
+
+	err = framework.AddToFrameworkScheme(monitoringv1.AddToScheme, &monitoringv1.ServiceMonitorList{})
+	Expect(err).ToNot(HaveOccurred())
+	err = framework.AddToFrameworkScheme(monitoringv1.AddToScheme, &monitoringv1.PrometheusRuleList{})
+	Expect(err).ToNot(HaveOccurred())
```
Use one line per check with `Exec(framework.AddToFrameworkSchema...).To(Succeed())`

RamLavi replied:
done

qinqon commented on Makefile:
```diff
@@ -121,6 +121,11 @@ docker-push-operator:
 docker-push-registry:
 	docker push $(IMAGE_REGISTRY)/$(REGISTRY_IMAGE):$(IMAGE_TAG)
 
+prom-rules-verify:
```
This are kind of integration tests and belong to ./test dir I think and have to be run as part of the integration tests

RamLavi replied:
well it's technically unit test.. I think it should be part of unit tests.
it's similar to what they do on [kubevirt](https://github.com/kubevirt/kubevirt/blob/main/Makefile#L167)

qinqon reviewed (changes requested):
Good work, it's a huge task, added some comments to make world a little better.

RamLavi commented on cmd/manager/main.go:
```diff
@@ -72,7 +61,7 @@ func main() {
 	mgr, err := manager.New(cfg, manager.Options{
 		Namespace:          namespace,
-		MetricsBindAddress: fmt.Sprintf("%s:%d", metricsHost, metricsPort),
+		MetricsBindAddress: "0",
```
I eventually moved it back in order to use conrtoller-runtime's prometheus endpint.

RamLavi commented on cluster/up.sh:
```diff
@@ -18,6 +18,13 @@ set -ex
+if [[ ($KUBEVIRT_PROVIDER =~ k8s-1\.1.*) || ($KUBEVIRT_PROVIDER =~ k8s-1.20) ]]; then
+	  echo "Prometheus is only supported for providers >= k8s-1.21\n Disabling Prometheus"
+    export KUBEVIRT_DEPLOY_PROMETHEUS=false
+fi
```
I tend to agree

oshoval replied:
I think kubevirtci doesn't want to deploy Prometheus by default ?

RamLavi commented on pkg/components/components.go:
```diff
@@ -179,6 +179,13 @@ func GetDeployment(version string, operatorVersion string, namespace string, rep
 							},
+							Ports: []corev1.ContainerPort{
+								corev1.ContainerPort{
+									Name:          "metrics",
+									Protocol:      "TCP",
+									ContainerPort: 8443,
```
didn't find const for TCP on http/net libraries.. do you know of any other places to look?

RamLavi commented on data/monitoring/prom-rule.yaml:
```diff
@@ -0,0 +1,11 @@
+spec:
+  groups:
+    - name: kubevirt.cnao.rules
+      rules:
```
no, fixed that.

RamLavi commented:
@qinqon fixes [here](https://github.com/kubevirt/cluster-network-addons-operator/compare/cacead5ccc2b09349d5ce311de4f84314d3e6339..80b49dd132735ebedd3986b521017f46ca871581)

RamLavi commented:
/retest

RamLavi commented:
/retest

oshoval commented on data/monitoring/prom-rule.yaml:
```diff
@@ -0,0 +1,20 @@
+          annotations:
+            summary: CNAO pod is down.
+          expr: kubevirt_cnao_num_up_operators == 0
+          for: 5m
```
aren't 5 minutes is too much ?
one might want to know about the alert faster

RamLavi replied:
I don't think there is cause to get it faster.
This is aligned with kubevirt's similar alert [VirtAPIDown](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-operator/resource/generate/components/crds.go#L564-L572)

erkanerol replied:
Please check https://github.com/kubevirt/kubevirt/issues/5383 and ensure that this alert works.

erkanerol replied:
@sradco Do we have a naming convention for alerts? Should it start with `Kubevirt`? See https://github.com/kubevirt/monitoring/tree/main/runbooks

sradco replied:
@RamLavi No naming convention, but you are missing runbook_url and a PR for the runbook in kubevirt/monitoring

RamLavi replied:
> Please check [kubevirt/kubevirt#5383](https://github.com/kubevirt/kubevirt/issues/5383) and ensure that this alert works.

I'm not using `labeldrop` so it shouldn't occur in this case. Also, I checked this on the GUI using CRC (see images in the PR description)

>@RamLavi No naming convention, but you are missing runbook_url and a PR for the runbook in kubevirt/monitoring

You are right. I will issue the relevant PRs for runbooks soon.

sradco replied:
@RamLavi The runbook_url should be part of this PR.

sradco replied:
Example of runbook_url https://github.com/kubevirt/hyperconverged-cluster-operator/pull/1448/files

oshoval commented on hack/prom-rule-ci/prom-rules-tests.yaml:
```diff
@@ -0,0 +1,21 @@
+
+tests:
+  - interval: 1m
+    input_series:
```
might need also a non zero vector for the tests?

RamLavi replied:
Do you mean a test checking that the alert doesn't fire on (for example) `"1 1 1 1 0 1"`?
@sradco is this possible on this unit test infra?

oshoval replied:
or does fire on `1 1 1 1 1` if needed

RamLavi replied:
it should only fire if it is 0 for more than 5 consecutive polls
Talked to shirly and there is a negative test scenario, will change, thanks!

RamLavi replied:
Done

oshoval commented on pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go:
```diff
@@ -593,6 +607,26 @@ func isSCCAvailable(c kubernetes.Interface) (bool, error) {
+		return true, nil
+	}
+
+	log.Printf("will not deploy monitoring manifests: not all monitoring resources are available: %s: %v, %s, %v", monitoringv1.PrometheusRuleKind, prometheusRuleResourceAvailable, monitoringv1.ServiceMonitorsKind, serviceMonitorResourceAvailable)
```
maybe worth to return this error string as err in the next line ?
this way the caller will print it as err, as it does for the other errors above

RamLavi replied:
It's kind of semantic, since I'm not really failing CNAO process if `IsMonitoringAvailable()` returns an error.
The difference between this printf and the other errors, is that in this case - nothing really failed. The cluster just doesn't have the right CRDs to run the monitoring instances, so we can't deploy monitoring. 
So instead of returning an error, I'm adding this printf that also sets it on the logs so it will be heard..

oshoval commented on pkg/monitoring/metrics.go:
```diff
@@ -0,0 +1,76 @@
+	return int32(portInt64)
+}
+
+func getNamespace() string {
+	monitoringNamespaceFromEnv := os.Getenv("MONITORING_NAMESPACE")
+
+	if monitoringNamespaceFromEnv != "" {
+		return monitoringNamespaceFromEnv
+	}
+	return defaultMonitoringNamespace
+}
+
+func getServiceAccount() string {
+	monitoringServiceAccountFromEnv := os.Getenv("MONITORING_SERVICE_ACCOUNT")
+
+	if monitoringServiceAccountFromEnv != "" {
+		return monitoringServiceAccountFromEnv
+	}
+	return defaultServiceAccountName
+}
```
maybe can be combined to one function which get an env var name + default value ?

RamLavi replied:
the fact that there is a value from env var or from default value is implementation detail, which we want to hide from the user using this API. it should just return one of them, preferably the one from the env var.

oshoval commented on pkg/network/network.go:
```diff
@@ -141,6 +143,13 @@ func Render(conf *cnao.NetworkAddonsConfigSpec, manifestDir string, openshiftNet
+	if err != nil {
+		return nil, err
+	}
+	objs = append(objs, o...)
```
maybe worth to do as an else, if there is err, the `o` might be partial theoretically, 
(not good practice to use a value when there is error)

RamLavi replied:
if an error s returned from a function, the value of the artifacts is not defined ( could be partial, could be nil, depends on implementation). it is up to the function calling to it to check if `err!= nil` and only otherwise use the returned variables.

oshoval replied:
Yes of course, this what i meant, just missed for a moment the `return`, because had in mind the monitoring is not mandatory

oshoval commented on test/check/check.go:
```diff
@@ -553,6 +591,51 @@ func checkForMutatingWebhookConfiguration(name string) error {
 	return nil
 }
 
+func checkForService(name string) error {
```
in the following functions, it makes the monitoring mandatory isn't it ?
opposed to the fact that you allow CNAO to be deployed even if the monitoring has an error

RamLavi replied:
The tests checking monitoring components should be skipped if the CRDs are not deployed..
I do skip the test checking it specifically if the CRDs are not deployed, but you might be right in the case of [AllComponents](https://github.com/kubevirt/cluster-network-addons-operator/pull/928/files#diff-dd3d0d54317e754fd95de914de01a73b5578626b42e755fdad4998186d552600R89)... hmmm.. let me see if I can dynamically remove it

oshoval replied:
Best lets inject an error that won't deploy monitoring and see what fails ? (unless you did that already)
it will what must be changed first

RamLavi replied:
I was thinking of wrapping `AllComponents` with `getAllComponents()` that will check and exclude monitoring if it is not deployed.. what do you think?

RamLavi replied:
So.. if we only keep the most updated provider, then I think we can keep `AllComponents`.
Since we always deploy prometheus, this the CRDs should always be deployed, and we should make sure that it acts as expected in our environment.
However, on production we don't want to force the user to deploy monitoring or CNAO will fail.

RamLavi commented on hack/prom-rule-ci/verify-rules.sh:
```diff
@@ -0,0 +1,68 @@
+source hack/components/yaml-utils.sh
+
+TMP_PATH=$(mktemp -d -p /tmp -t alerts-test-XXXXXX)
+PROMETHERUS_IMAGE=docker.io/prom/prometheus:v2.15.2
```
@erkanerol is there a quay mirror for this?

erkanerol replied:
I think it is here https://quay.io/repository/prometheus/prometheus

RamLavi commented on hack/prom-rule-ci/verify-rules.sh:
```diff
@@ -0,0 +1,68 @@
+  local tmp_prom_spec=${TMP_PATH}/prom-spec
+  local tmp_tests_file=${TMP_PATH}/prom-rules-tests.yaml
+
+  trap teardown EXIT SIGINT SIGTERM
```
I don't see it running twice (from line [8677](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_cluster-network-addons-operator/928/pull-cluster-network-addons-operator-unit-test/1420404095063166976#1:build-log.txt%3A8677) to line [8801](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_cluster-network-addons-operator/928/pull-cluster-network-addons-operator-unit-test/1420404095063166976#1:build-log.txt%3A8801))

oshoval replied:
It depends on the signal you received,
if you will get SIGTERM from prow, for example when you force push, this what what it does
it will run both the SIGTERM and the EXIT, we saw it in the past on kubevirtci, 
anyhow imo SIGTERM is redundant

RamLavi replied:
Done

RamLavi commented on pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go:
```diff
@@ -91,6 +92,14 @@ func Add(mgr manager.Manager) error {
+	if err != nil {
+		// we don't want CNAO to fail only if monitoring cannot be activated.
+		addMonitorServiceResources = false
+		fmt.Printf("failed to check for availability of Monitoring namespace: %v", err)
```
it could be nice to have, but since statusManager (and correspondingly the eventEmitter) are only initialized when the reconciler start to work and this code runs before that, then I'm not sure it's worth the trouble.

RamLavi commented on pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go:
```diff
@@ -504,6 +513,11 @@ func updateObjectsLabels(crLables map[string]string, objs []*unstructured.Unstru
 		// Label objects with version of the operator they were created by
 		labels[cnaov1.SchemeGroupVersion.Group+"/version"] = operatorVersionLabel
+		labels[names.PROMETHEUS_LABEL_KEY] = ""
+		err = updateObjectTemplateLabels(obj, labels, []string{names.PROMETHEUS_LABEL_KEY})
+		if err != nil {
+			return err
+		}
```
`appLabelKeys` serves both `updateLabelsFromCR` and `updateObjectTemplateLabels` and `PROMETHEUS_LABEL_KEY` doesn't relate to the former function (the value is not fetched from the CR - but rather is always `""`) So I deemed them separate cases and used it on its own merit

oshoval commented on test/check/components.go:
```diff
@@ -76,13 +80,20 @@ var (
+	MonitoringComponent = Component{
+		ComponentName:  "Monitoring",
+		Service:        "cluster-network-addons-operator-prometheus-metrics",
+		ServiceMonitor: "prometheus-rules-cluster-network-addons-operator",
```
those two lines should be same value?
ServiceMonitor, PrometheusRule

RamLavi replied:
I mainly did it cause kubevirt's [serviceMonitor](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-operator/resource/generate/components/crds.go#L496) and [prometheusRule](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-operator/resource/generate/components/crds.go#L539) have the same name, but It's not really mandatory and I haven't seen it done in other projects...
so.. do you think I should change it?

oshoval replied:
maybe call them `prometheus-cnao-service-monitor`
and
`prometheus-cnao-rule` ?

RamLavi replied:
I think we should keep it consistent, we don't use cnao abbrev on CNAO instances, 
so I think we should only change the `ServiceMonitor` to `service-monitor-cluster-network-addons-operator`

oshoval commented on test/e2e/workflow/deployment_test.go:
```diff
@@ -89,6 +90,14 @@ var _ = Describe("NetworkAddonsConfig", func() {
+		It("should deploy prometheus if NetworkAddonsConfigSpec is not empty", func() {
+			isMonitoringDeployed, err := networkaddonsconfig.IsMonitoringAvailable(framework.Global.KubeClient)
+			Expect(err).ToNot(HaveOccurred())
+			if !isMonitoringDeployed {
```
why allowing in the tests to be not deployed and skip ?

RamLavi replied:
well it depends.. This logic is simply mimicking what we do in the production code.
But what it means in the context of this tests is that if we don't skip this test then k8s providers <1.21 will just fail this test/e2e/workflow lane.
Do you think we should not care of old providers?

oshoval replied:
ye lets not care about them
if and only if we will need it in the future we can reconsider, but i bet we won't need it
so better to be strict + less to maintain

RamLavi replied:
I agree. Done

RamLavi commented:
/retest
seems unrelated

qinqon commented:
/lgtm

sradco commented on data/monitoring/prom-rule.yaml:
```diff
@@ -0,0 +1,20 @@
+          expr: kubevirt_cnao_num_up_operators == 0
+          for: 5m
+          labels:
+            severity: critical
```
@RamLavi  Does this alert severity fit the description?
Critical - When the service is down and you loss critical functionality, an action is required immediately

RamLavi replied:
if CNAO operator is down then there's no one to make sure that CNAO's components are running, one of them is Kubemacpool. if Kubemacpool is down, then VMs can't be created. 
so considering your definition, maybe warning is more fitting?

RamLavi replied:
changed to Warning. Done

oshoval commented on test/e2e/workflow/deployment_test.go:
```diff
@@ -89,6 +89,9 @@ var _ = Describe("NetworkAddonsConfig", func() {
 			),
 		)
+		It("should deploy prometheus if NetworkAddonsConfigSpec is not empty", func() {
+			testConfigCreate(gvk, cnao.NetworkAddonsConfigSpec{MacvtapCni: &cnao.MacvtapCni{}}, []Component{MacvtapComponent, MonitoringComponent})
```
why actually not to deploy prometheus even if spec is empty ?
some of the gauges might still need to be available aren't they ?

RamLavi replied:
it's simply the way CNAO deploys components, it only does it if the CR is there and not empty.

oshoval commented on data/monitoring/service.yaml:
```diff
@@ -0,0 +1,17 @@
+spec:
+  ports:
+    - name: metrics
+      port: 8080
```
should this be a variable, that you set via the Render function in case the port is not the default one ?

RamLavi replied:
it does not.. I'm exposing port internal 8080 to external port 8080. the internal is already defined as port `metrics` (in the next line). this 8080 is actually a free number that I chose to be the same to avoid future confusion..

oshoval commented:
lgtm from me, thanks,
just letting others review as well

qinqon commented:
/lgtm
/approve
/restest

qinqon commented:
/retest

qinqon commented:
/hold

RamLavi commented:
> /hold
> 
> Some answers missing.

yes let's also wait on @sradco lgtm, I'm currently waiting for the new runbook url format.

RamLavi commented:
@sradco Added runbook URL [here](https://github.com/kubevirt/cluster-network-addons-operator/compare/b035e4aac16df185243d35005a596cf1cb9086f9..0bdee788e60f5aabeef99004d1423b4a5707dc6c)

RamLavi commented:
removed the prefix [here](https://github.com/kubevirt/cluster-network-addons-operator/compare/0bdee788e60f5aabeef99004d1423b4a5707dc6c..22b60f25bda44fe9260c98a0df62e87d0f2026cd), as it messes with the prometheus unit-test module
/retest

sradco commented:
/lgtm

sradco commented:
/lgtm

oshoval commented:
To remove the hold?

RamLavi commented:
> To remove the hold?

yeah

/hold cancel
